File: app/src/main/java/com/senderlink/app/model/GroupMessage.kt
package com.senderlink.app.model

import com.google.gson.annotations.SerializedName

/**
 * Modelo de mensaje de chat grupal
 * Mapea la respuesta del backend MongoDB
 */
data class GroupMessage(
    @SerializedName("_id")
    val id: String,

    @SerializedName("chatId")
    val chatId: String,

    @SerializedName("senderUid")
    val senderUid: String,

    @SerializedName("senderName")
    val senderName: String,

    @SerializedName("senderPhoto")
    val senderPhoto: String? = null,

    @SerializedName("text")
    val text: String,

    @SerializedName("type")
    val type: String = "TEXT",

    @SerializedName("createdAt")
    val createdAt: String,

    @SerializedName("updatedAt")
    val updatedAt: String
) {
    fun isMine(currentUid: String): Boolean = senderUid == currentUid

    fun isSystemMessage(): Boolean = type == "SYSTEM"

    fun getFormattedTime(): String {
        return try {
            createdAt.substring(11, 16) // "2026-01-21T14:30:00.000Z" → "14:30"
        } catch (e: Exception) {
            "00:00"
        }
    }

    fun hasSenderPhoto(): Boolean {
        return !senderPhoto.isNullOrBlank() &&
                senderPhoto != "null" &&
                senderPhoto != ""
    }
}

File: senderlink_api/models/GroupMessage.js
const mongoose = require("mongoose");

/**
 * ===========================================
 * MODELO GROUP MESSAGE (Mensajes Grupales)
 * ===========================================
 *
 * Un mensaje grupal es un mensaje enviado dentro
 * del chat de un evento grupal.
 *
 * DIFERENCIAS CON Message.js (chat 1-a-1):
 * - No tiene destinatarioUid (es para todo el grupo)
 * - No tiene campo "leido" (todos leen el mismo mensaje)
 * - Incluye nombre y foto del remitente (para mostrar en UI)
 */

const GroupMessageSchema = new mongoose.Schema(
  {
    // 💬 ID DEL CHAT GRUPAL
    // Este chatId viene del campo chatId de EventoGrupal
    chatId: {
      type: String,
      required: true,
      index: true // Importante para consultas rápidas
    },

    // 👤 REMITENTE (quien envía el mensaje)
    senderUid: {
      type: String,
      required: true,
      index: true
    },
    senderName: {
      type: String,
      required: true
    },
    senderPhoto: {
      type: String,
      default: ""
    },

    // 📝 CONTENIDO DEL MENSAJE
    text: {
      type: String,
      required: true,
      trim: true,
      maxlength: 500 // Límite de caracteres
    },

    // 🏷️ TIPO DE MENSAJE (para futuras expansiones)
    type: {
      type: String,
      enum: ["TEXT", "SYSTEM", "IMAGE"],
      default: "TEXT"
    }
  },
  {
    timestamps: true // Crea automáticamente createdAt y updatedAt
  }
);

// ===========================================
// ÍNDICES PARA CONSULTAS EFICIENTES
// ===========================================

// Índice compuesto: permite buscar mensajes de un chat
// ordenados por fecha de forma muy rápida
GroupMessageSchema.index({ chatId: 1, createdAt: -1 });

// ===========================================
// EXPORTAR MODELO
// ===========================================

module.exports = mongoose.model("GroupMessage", GroupMessageSchema);